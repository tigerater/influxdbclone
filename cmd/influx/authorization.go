package main

import (
	"context"
	"os"

	platform "github.com/influxdata/influxdb"
	"github.com/influxdata/influxdb/cmd/influx/internal"
	"github.com/influxdata/influxdb/http"
	"github.com/spf13/cobra"
)

var authorizationCmd = &cobra.Command{
	Use:     "auth",
	Aliases: []string{"authorization"},
	Short:   "Authorization management commands",
	Run: func(cmd *cobra.Command, args []string) {
		cmd.Usage()
	},
}

// AuthorizationCreateFlags are command line args used when creating a authorization
type AuthorizationCreateFlags struct {
	user string
	org  string

	writeUserPermission bool
	readUserPermission  bool

	writeBucketsPermission bool
	readBucketsPermission  bool

	writeBucketPermissions []string
	readBucketPermissions  []string

	writeTasksPermission bool
	readTasksPermission  bool

	writeTelegrafsPermission bool
	readTelegrafsPermission  bool

	writeOrganizationsPermission bool
	readOrganizationsPermission  bool

	writeDashboardsPermission bool
	readDashboardsPermission  bool
}

var authorizationCreateFlags AuthorizationCreateFlags

func init() {
	authorizationCreateCmd := &cobra.Command{
		Use:   "create",
		Short: "Create authorization",
		RunE:  wrapCheckSetup(authorizationCreateF),
	}

	authorizationCreateCmd.Flags().StringVarP(&authorizationCreateFlags.org, "org", "o", "", "The organization name (required)")
	authorizationCreateCmd.MarkFlagRequired("org")

	authorizationCreateCmd.Flags().StringVarP(&authorizationCreateFlags.user, "user", "u", "", "The user name")

	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.writeUserPermission, "write-user", "", false, "Grants the permission to perform mutative actions against organization users")
	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.readUserPermission, "read-user", "", false, "Grants the permission to perform read actions against organization users")

	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.writeBucketsPermission, "write-buckets", "", false, "Grants the permission to perform mutative actions against organization buckets")
	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.readBucketsPermission, "read-buckets", "", false, "Grants the permission to perform read actions against organization buckets")

	authorizationCreateCmd.Flags().StringArrayVarP(&authorizationCreateFlags.writeBucketPermissions, "write-bucket", "", []string{}, "The bucket id")
	authorizationCreateCmd.Flags().StringArrayVarP(&authorizationCreateFlags.readBucketPermissions, "read-bucket", "", []string{}, "The bucket id")

	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.writeTasksPermission, "write-tasks", "", false, "Grants the permission to create tasks")
	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.readTasksPermission, "read-tasks", "", false, "Grants the permission to read tasks")

	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.writeTelegrafsPermission, "write-telegrafs", "", false, "Grants the permission to create telegraf configs")
	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.readTelegrafsPermission, "read-telegrafs", "", false, "Grants the permission to read telegraf configs")

	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.writeOrganizationsPermission, "write-orgs", "", false, "Grants the permission to create organizations")
	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.readOrganizationsPermission, "read-orgs", "", false, "Grants the permission to read organizations")

	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.writeDashboardsPermission, "write-dashboards", "", false, "Grants the permission to create dashboards")
	authorizationCreateCmd.Flags().BoolVarP(&authorizationCreateFlags.readDashboardsPermission, "read-dashboards", "", false, "Grants the permission to read dashboards")

	authorizationCmd.AddCommand(authorizationCreateCmd)
}

func authorizationCreateF(cmd *cobra.Command, args []string) error {
	var permissions []platform.Permission
	orgSvc, err := newOrganizationService(flags)
	if err != nil {
		return err
	}

	ctx := context.Background()
	orgFilter := platform.OrganizationFilter{Name: &authorizationCreateFlags.org}
	o, err := orgSvc.FindOrganization(ctx, orgFilter)
	if err != nil {
		return err
	}

	if authorizationCreateFlags.writeUserPermission {
		p, err := platform.NewPermission(platform.WriteAction, platform.UsersResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	if authorizationCreateFlags.readUserPermission {
		p, err := platform.NewPermission(platform.ReadAction, platform.UsersResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	if authorizationCreateFlags.writeBucketsPermission {
		p, err := platform.NewPermission(platform.WriteAction, platform.BucketsResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	if authorizationCreateFlags.readBucketsPermission {
		p, err := platform.NewPermission(platform.ReadAction, platform.BucketsResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	for _, p := range authorizationCreateFlags.writeBucketPermissions {
		var id platform.ID
		if err := id.DecodeFromString(p); err != nil {
			return err
		}

		p, err := platform.NewPermissionAtID(id, platform.WriteAction, platform.BucketsResourceType, o.ID)
		if err != nil {
			return err
		}

		permissions = append(permissions, *p)
	}

	for _, p := range authorizationCreateFlags.readBucketPermissions {
		var id platform.ID
		if err := id.DecodeFromString(p); err != nil {
			return err
		}

		p, err := platform.NewPermissionAtID(id, platform.ReadAction, platform.BucketsResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	if authorizationCreateFlags.writeTasksPermission {
		p, err := platform.NewPermission(platform.WriteAction, platform.TasksResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	if authorizationCreateFlags.readTasksPermission {
		p, err := platform.NewPermission(platform.ReadAction, platform.TasksResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	if authorizationCreateFlags.writeTelegrafsPermission {
		p, err := platform.NewPermission(platform.WriteAction, platform.TelegrafsResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	if authorizationCreateFlags.readTelegrafsPermission {
		p, err := platform.NewPermission(platform.ReadAction, platform.TelegrafsResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	if authorizationCreateFlags.writeOrganizationsPermission {
		p, err := platform.NewPermission(platform.WriteAction, platform.OrgsResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	if authorizationCreateFlags.readOrganizationsPermission {
		p, err := platform.NewPermission(platform.ReadAction, platform.OrgsResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	if authorizationCreateFlags.writeDashboardsPermission {
		p, err := platform.NewPermission(platform.WriteAction, platform.DashboardsResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	if authorizationCreateFlags.readDashboardsPermission {
		p, err := platform.NewPermission(platform.ReadAction, platform.DashboardsResourceType, o.ID)
		if err != nil {
			return err
		}
		permissions = append(permissions, *p)
	}

	authorization := &platform.Authorization{
		Permissions: permissions,
		OrgID:       o.ID,
	}

	if userName := authorizationCreateFlags.user; userName != "" {
		userSvc, err := newUserService(flags)
		if err != nil {
			return err
		}
		user, err := userSvc.FindUser(ctx, platform.UserFilter{
			Name: &userName,
		})
		if err != nil {
			return err
		}
		authorization.UserID = user.ID
	}

	s, err := newAuthorizationService(flags)
	if err != nil {
		return err
	}

	if err := s.CreateAuthorization(context.Background(), authorization); err != nil {
		return err
	}

	w := internal.NewTabWriter(os.Stdout)
	w.WriteHeaders(
		"ID",
		"Token",
		"Status",
		"UserID",
		"Permissions",
	)

	ps := []string{}
	for _, p := range authorization.Permissions {
		ps = append(ps, p.String())
	}

	w.Write(map[string]interface{}{
		"ID":          authorization.ID.String(),
		"Token":       authorization.Token,
		"Status":      authorization.Status,
		"UserID":      authorization.UserID.String(),
		"Permissions": ps,
	})

	w.Flush()

	return nil
}

// AuthorizationFindFlags are command line args used when finding a authorization
type AuthorizationFindFlags struct {
	user   string
	userID string
	org    string
	orgID  string
	id     string
}

var authorizationFindFlags AuthorizationFindFlags

func init() {
	authorizationFindCmd := &cobra.Command{
		Use:   "find",
		Short: "Find authorization",
		RunE:  wrapCheckSetup(authorizationFindF),
	}

	authorizationFindCmd.Flags().StringVarP(&authorizationFindFlags.user, "user", "u", "", "The user")
	authorizationFindCmd.Flags().StringVarP(&authorizationFindFlags.userID, "user-id", "", "", "The user ID")
	authorizationFindCmd.Flags().StringVarP(&authorizationFindFlags.org, "org", "o", "", "The org")
	authorizationFindCmd.Flags().StringVarP(&authorizationFindFlags.orgID, "org-id", "", "", "The org ID")
	authorizationFindCmd.Flags().StringVarP(&authorizationFindFlags.id, "id", "i", "", "The authorization ID")

	authorizationCmd.AddCommand(authorizationFindCmd)
}

func newAuthorizationService(f Flags) (platform.AuthorizationService, error) {
	if flags.local {
		return newLocalKVService()
	}
	return &http.AuthorizationService{
		Addr:  flags.host,
		Token: flags.token,
	}, nil
}

func authorizationFindF(cmd *cobra.Command, args []string) error {
	s, err := newAuthorizationService(flags)
	if err != nil {
		return err
	}

	filter := platform.AuthorizationFilter{}
	if authorizationFindFlags.id != "" {
		fID, err := platform.IDFromString(authorizationFindFlags.id)
		if err != nil {
			return err
		}
		filter.ID = fID
	}
	if authorizationFindFlags.user != "" {
		filter.User = &authorizationFindFlags.user
	}
	if authorizationFindFlags.userID != "" {
		uID, err := platform.IDFromString(authorizationFindFlags.userID)
		if err != nil {
			return err
		}
		filter.UserID = uID
	}
	if authorizationFindFlags.org != "" {
		filter.Org = &authorizationFindFlags.org
	}
	if authorizationFindFlags.orgID != "" {
		oID, err := platform.IDFromString(authorizationFindFlags.orgID)
		if err != nil {
			return err
		}
		filter.OrgID = oID
	}

	authorizations, _, err := s.FindAuthorizations(context.Background(), filter)
	if err != nil {
		return err
	}

	w := internal.NewTabWriter(os.Stdout)
	w.WriteHeaders(
		"ID",
		"Token",
		"Status",
		"User",
		"UserID",
		"Permissions",
	)

	for _, a := range authorizations {
		var permissions []string
		for _, p := range a.Permissions {
			permissions = append(permissions, p.String())
		}

		w.Write(map[string]interface{}{
			"ID":          a.ID,
			"Token":       a.Token,
			"Status":      a.Status,
			"UserID":      a.UserID.String(),
			"Permissions": permissions,
		})
	}

	w.Flush()

	return nil
}

// AuthorizationDeleteFlags are command line args used when deleting a authorization
type AuthorizationDeleteFlags struct {
	id string
}

var authorizationDeleteFlags AuthorizationDeleteFlags

func init() {
	authorizationDeleteCmd := &cobra.Command{
		Use:   "delete",
		Short: "Delete authorization",
		RunE:  wrapCheckSetup(authorizationDeleteF),
	}

	authorizationDeleteCmd.Flags().StringVarP(&authorizationDeleteFlags.id, "id", "i", "", "The authorization ID (required)")
	authorizationDeleteCmd.MarkFlagRequired("id")

	authorizationCmd.AddCommand(authorizationDeleteCmd)
}

func authorizationDeleteF(cmd *cobra.Command, args []string) error {
	s, err := newAuthorizationService(flags)
	if err != nil {
		return err
	}

	id, err := platform.IDFromString(authorizationDeleteFlags.id)
	if err != nil {
		return err
	}

	ctx := context.TODO()
	a, err := s.FindAuthorizationByID(ctx, *id)
	if err != nil {
		return err
	}

	if err := s.DeleteAuthorization(context.Background(), *id); err != nil {
		return err
	}

	w := internal.NewTabWriter(os.Stdout)
	w.WriteHeaders(
		"ID",
		"Token",
		"User",
		"UserID",
		"Permissions",
		"Deleted",
	)

	ps := []string{}
	for _, p := range a.Permissions {
		ps = append(ps, p.String())
	}

	w.Write(map[string]interface{}{
		"ID":          a.ID.String(),
		"Token":       a.Token,
		"UserID":      a.UserID.String(),
		"Permissions": ps,
		"Deleted":     true,
	})

	w.Flush()

	return nil
}

// AuthorizationActiveFlags are command line args used when enabling an authorization
type AuthorizationActiveFlags struct {
	id string
}

var authorizationActiveFlags AuthorizationActiveFlags

func init() {
	authorizationActiveCmd := &cobra.Command{
		Use:   "active",
		Short: "Active authorization",
		RunE:  wrapCheckSetup(authorizationActiveF),
	}

	authorizationActiveCmd.Flags().StringVarP(&authorizationActiveFlags.id, "id", "i", "", "The authorization ID (required)")
	authorizationActiveCmd.MarkFlagRequired("id")

	authorizationCmd.AddCommand(authorizationActiveCmd)
}

func authorizationActiveF(cmd *cobra.Command, args []string) error {
	s, err := newAuthorizationService(flags)
	if err != nil {
		return err
	}

	var id platform.ID
	if err := id.DecodeFromString(authorizationActiveFlags.id); err != nil {
		return err
	}

	ctx := context.TODO()
	if _, err := s.FindAuthorizationByID(ctx, id); err != nil {
		return err
	}

	a, err := s.UpdateAuthorization(context.Background(), id, &platform.AuthorizationUpdate{
		Status: platform.Active.Ptr(),
	})
	if err != nil {
		return err
	}

	w := internal.NewTabWriter(os.Stdout)
	w.WriteHeaders(
		"ID",
		"Token",
		"Status",
		"User",
		"UserID",
		"Permissions",
	)

	ps := []string{}
	for _, p := range a.Permissions {
		ps = append(ps, p.String())
	}

	w.Write(map[string]interface{}{
		"ID":          a.ID.String(),
		"Token":       a.Token,
		"Status":      a.Status,
		"UserID":      a.UserID.String(),
		"Permissions": ps,
	})

	w.Flush()

	return nil
}

// AuthorizationInactiveFlags are command line args used when disabling an authorization
type AuthorizationInactiveFlags struct {
	id string
}

var authorizationInactiveFlags AuthorizationInactiveFlags

func init() {
	authorizationInactiveCmd := &cobra.Command{
		Use:   "inactive",
		Short: "Inactive authorization",
		RunE:  wrapCheckSetup(authorizationInactiveF),
	}

	authorizationInactiveCmd.Flags().StringVarP(&authorizationInactiveFlags.id, "id", "i", "", "The authorization ID (required)")
	authorizationInactiveCmd.MarkFlagRequired("id")

	authorizationCmd.AddCommand(authorizationInactiveCmd)
}

func authorizationInactiveF(cmd *cobra.Command, args []string) error {
	s, err := newAuthorizationService(flags)
	if err != nil {
		return err
	}

	var id platform.ID
	if err := id.DecodeFromString(authorizationInactiveFlags.id); err != nil {
		return err
	}

	ctx := context.TODO()
	if _, err = s.FindAuthorizationByID(ctx, id); err != nil {
		return err
	}

	a, err := s.UpdateAuthorization(context.Background(), id, &platform.AuthorizationUpdate{
		Status: platform.Inactive.Ptr(),
	})
	if err != nil {
		return err
	}

	w := internal.NewTabWriter(os.Stdout)
	w.WriteHeaders(
		"ID",
		"Token",
		"Status",
		"User",
		"UserID",
		"Permissions",
	)

	ps := []string{}
	for _, p := range a.Permissions {
		ps = append(ps, p.String())
	}

	w.Write(map[string]interface{}{
		"ID":          a.ID.String(),
		"Token":       a.Token,
		"Status":      a.Status,
		"UserID":      a.UserID.String(),
		"Permissions": ps,
	})

	w.Flush()

	return nil
}
