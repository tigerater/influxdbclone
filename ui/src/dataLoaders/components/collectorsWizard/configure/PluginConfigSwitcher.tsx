// Libraries
import React, {PureComponent} from 'react'
import {connect} from 'react-redux'
import _ from 'lodash'

// Components
import PluginConfigForm from 'src/dataLoaders/components/collectorsWizard/configure/PluginConfigForm'
import EmptyDataSourceState from 'src/dataLoaders/components/configureStep/EmptyDataSourceState'

// Utils
import {getConfigFields} from 'src/dataLoaders/utils/pluginConfigs'

// Types
import {TelegrafPlugin, ConfigFields} from 'src/types/dataLoaders'
import {AppState} from 'src/types'
import TelegrafPluginInstructions from 'src/dataLoaders/components/collectorsWizard/configure/TelegrafPluginInstructions'

interface StateProps {
  telegrafPlugins: TelegrafPlugin[]
}

type Props = StateProps

export class PluginConfigSwitcher extends PureComponent<Props> {
  public render() {
    const {telegrafPlugins} = this.props

    if (this.activeTelegrafPlugin) {
      return (
        <PluginConfigForm
          telegrafPlugin={this.activeTelegrafPlugin}
          configFields={this.configFields}
        />
      )
    } else if (!telegrafPlugins || !telegrafPlugins.length) {
      return <EmptyDataSourceState />
    }

    return <TelegrafPluginInstructions />
  }

  private get activeTelegrafPlugin(): TelegrafPlugin {
    const {telegrafPlugins} = this.props
    return telegrafPlugins.find(tp => tp.active)
  }

  private get configFields(): ConfigFields {
    if (this.activeTelegrafPlugin) {
      return getConfigFields(this.activeTelegrafPlugin.name)
    }
  }
}

const mstp = ({
  dataLoading: {
    dataLoaders: {telegrafPlugins},
  },
}: AppState): StateProps => ({
  telegrafPlugins,
})

export default connect<StateProps>(
  mstp,
  null
)(PluginConfigSwitcher)
