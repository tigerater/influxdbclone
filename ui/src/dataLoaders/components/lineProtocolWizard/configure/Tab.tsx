import React, {PureComponent} from 'react'

import {Radio} from '@influxdata/clockface'

import {LineProtocolTab} from 'src/types'

interface Props {
  active: boolean
  tab: LineProtocolTab
  onClick: (tab: LineProtocolTab) => void
}

export default class extends PureComponent<Props> {
  public render() {
    const {tab, active} = this.props

    return (
      <Radio.Button
        key={tab}
        id={tab}
        titleText={tab}
        value={tab}
        active={active}
        onClick={this.handleClick}
      >
        {tab}
      </Radio.Button>
    )
  }

  private handleClick = () => {
    this.props.onClick(this.props.tab)
  }
}
