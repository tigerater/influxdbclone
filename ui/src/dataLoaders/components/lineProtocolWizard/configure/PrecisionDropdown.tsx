// Libraries
import React, {PureComponent} from 'react'

// Components
import {Dropdown} from '@influxdata/clockface'

// Types
import {WritePrecision} from '@influxdata/influx'
import {Precision} from 'src/types/dataLoaders'

interface Props {
  setPrecision: (precision: WritePrecision) => void
  precision: WritePrecision
}

const writePrecisions: WritePrecision[] = [
  WritePrecision.Ns,
  WritePrecision.Us,
  WritePrecision.Ms,
  WritePrecision.S,
]

const makePrecisionReadable = {
  [WritePrecision.Ns]: Precision.Nanoseconds,
  [WritePrecision.Us]: Precision.Microseconds,
  [WritePrecision.S]: Precision.Seconds,
  [WritePrecision.Ms]: Precision.Milliseconds,
}

class PrecisionDropdown extends PureComponent<Props> {
  public render() {
    const {setPrecision, precision} = this.props
    return (
      <div className="wizard-step--lp-precision">
        <label>Time Precision</label>
        <Dropdown
          widthPixels={200}
          className="wizard-step--lp-precision"
          button={(active, onClick) => (
            <Dropdown.Button active={active} onClick={onClick}>
              {makePrecisionReadable[precision]}
            </Dropdown.Button>
          )}
          menu={onCollapse => (
            <Dropdown.Menu onCollapse={onCollapse}>
              {writePrecisions.map(value => (
                <Dropdown.Item
                  key={value}
                  value={value}
                  id={value}
                  onClick={setPrecision}
                  selected={`${value}` === `${precision}`}
                >
                  {makePrecisionReadable[value]}
                </Dropdown.Item>
              ))}
            </Dropdown.Menu>
          )}
        />
      </div>
    )
  }
}

export default PrecisionDropdown
