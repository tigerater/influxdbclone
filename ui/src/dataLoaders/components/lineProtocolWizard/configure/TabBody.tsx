// Libraries
import React, {PureComponent, ChangeEvent} from 'react'

// Components
import {
  Form,
  Input,
  Grid,
  TextArea,
  Columns,
  InputType,
  ComponentSize,
} from '@influxdata/clockface'
import DragAndDrop from 'src/shared/components/DragAndDrop'
import {LineProtocolTab} from 'src/types'
import {WritePrecision} from '@influxdata/influx'

import {setLineProtocolBody} from 'src/dataLoaders/actions/dataLoaders'

interface Props {
  lineProtocolBody: string
  activeLPTab: LineProtocolTab
  precision: WritePrecision
  setLineProtocolBody: typeof setLineProtocolBody
  onURLChange: (url: string) => void
  urlInput: string
}

export default class extends PureComponent<Props> {
  public render() {
    const {lineProtocolBody, activeLPTab, urlInput} = this.props

    switch (activeLPTab) {
      case LineProtocolTab.UploadFile:
        return (
          <DragAndDrop
            submitText="Upload File"
            handleSubmit={this.handleSetLineProtocol}
            submitOnDrop={true}
            submitOnUpload={true}
          />
        )
      case LineProtocolTab.EnterManually:
        return (
          <TextArea
            value={lineProtocolBody}
            placeholder="Write text here"
            onChange={this.handleTextChange}
          />
        )
      case LineProtocolTab.EnterURL:
        return (
          <Grid>
            <Grid.Row>
              <Grid.Column
                widthXS={Columns.Twelve}
                widthMD={Columns.Ten}
                offsetMD={Columns.One}
              >
                <Form.Element label="File URL:">
                  <Input
                    titleText="File URL:"
                    type={InputType.Text}
                    placeholder="http://..."
                    value={urlInput}
                    onChange={this.handleChange}
                    autoFocus={true}
                    size={ComponentSize.Large}
                  />
                </Form.Element>
              </Grid.Column>
            </Grid.Row>
          </Grid>
        )
    }
  }

  private handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const {value} = e.target
    this.props.onURLChange(value)
  }

  private handleTextChange = async (e: ChangeEvent<HTMLTextAreaElement>) => {
    const {setLineProtocolBody} = this.props
    setLineProtocolBody(e.target.value)
  }

  private handleSetLineProtocol = (lpBody: string) => {
    const {setLineProtocolBody} = this.props
    setLineProtocolBody(lpBody)
  }
}
