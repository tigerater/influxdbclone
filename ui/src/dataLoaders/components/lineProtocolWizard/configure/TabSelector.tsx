import React, {PureComponent} from 'react'

import {Radio, ButtonShape} from '@influxdata/clockface'
import {LineProtocolTab} from 'src/types'

import Tab from 'src/dataLoaders/components/lineProtocolWizard/configure/Tab'

interface Props {
  tabs: LineProtocolTab[]
  activeLPTab: LineProtocolTab
  onClick: (tab: LineProtocolTab) => void
}

export default class extends PureComponent<Props> {
  public render() {
    const {tabs, activeLPTab} = this.props
    return (
      <Radio shape={ButtonShape.Default}>
        {tabs.map(t => (
          <Tab
            tab={t}
            key={t}
            active={activeLPTab === t}
            onClick={this.handleTabClick}
          />
        ))}
      </Radio>
    )
  }

  private handleTabClick = (tab: LineProtocolTab) => {
    const {activeLPTab, onClick} = this.props
    if (tab !== activeLPTab) {
      onClick(tab)
    }
  }
}
