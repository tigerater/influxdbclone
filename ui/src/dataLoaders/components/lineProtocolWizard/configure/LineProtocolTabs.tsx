// Libraries
import React, {PureComponent} from 'react'
import {connect} from 'react-redux'
import _ from 'lodash'

// Components
import PrecisionDropdown from 'src/dataLoaders/components/lineProtocolWizard/configure/PrecisionDropdown'
import TabSelector from 'src/dataLoaders/components/lineProtocolWizard/configure/TabSelector'
import TabBody from 'src/dataLoaders/components/lineProtocolWizard/configure/TabBody'

// Types
import {AppState, LineProtocolTab} from 'src/types'
import {WritePrecision} from '@influxdata/influx'

// Actions
import {
  setLineProtocolBody,
  setActiveLPTab,
  setPrecision,
} from 'src/dataLoaders/actions/dataLoaders'

interface OwnProps {
  tabs: LineProtocolTab[]
  bucket: string
  org: string
}

type Props = OwnProps & DispatchProps & StateProps

interface DispatchProps {
  setLineProtocolBody: typeof setLineProtocolBody
  setActiveLPTab: typeof setActiveLPTab
  setPrecision: typeof setPrecision
}

interface StateProps {
  lineProtocolBody: string
  activeLPTab: LineProtocolTab
  precision: WritePrecision
}

interface State {
  urlInput: string
}

export class LineProtocolTabs extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      urlInput: '',
    }
  }

  public render() {
    const {
      setPrecision,
      precision,
      activeLPTab,
      tabs,
      setLineProtocolBody,
      lineProtocolBody,
    } = this.props

    const {urlInput} = this.state

    return (
      <>
        <div className="line-protocol--header">
          <TabSelector
            activeLPTab={activeLPTab}
            tabs={tabs}
            onClick={this.handleTabClick}
          />
          <PrecisionDropdown
            setPrecision={setPrecision}
            precision={precision}
          />
        </div>
        <TabBody
          onURLChange={this.handleURLChange}
          activeLPTab={activeLPTab}
          precision={precision}
          urlInput={urlInput}
          lineProtocolBody={lineProtocolBody}
          setLineProtocolBody={setLineProtocolBody}
        />
      </>
    )
  }

  private handleTabClick = (tab: LineProtocolTab) => {
    const {setActiveLPTab, setLineProtocolBody} = this.props

    setLineProtocolBody('')
    setActiveLPTab(tab)
  }

  private handleURLChange = (urlInput: string) => {
    this.setState({urlInput})
  }
}

const mstp = ({
  dataLoading: {
    dataLoaders: {lineProtocolBody, activeLPTab, precision},
  },
}: AppState) => {
  return {lineProtocolBody, activeLPTab, precision}
}

const mdtp: DispatchProps = {
  setLineProtocolBody,
  setActiveLPTab,
  setPrecision,
}

export default connect<StateProps, DispatchProps, OwnProps>(
  mstp,
  mdtp
)(LineProtocolTabs)
