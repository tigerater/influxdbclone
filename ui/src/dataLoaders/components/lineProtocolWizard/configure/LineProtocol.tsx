// Libraries
import React, {PureComponent} from 'react'
import {connect} from 'react-redux'

import _ from 'lodash'

// Components
import {Form} from '@influxdata/clockface'
import LineProtocolTabs from 'src/dataLoaders/components/lineProtocolWizard/configure/LineProtocolTabs'
import OnboardingButtons from 'src/onboarding/components/OnboardingButtons'
import FancyScrollbar from 'src/shared/components/fancy_scrollbar/FancyScrollbar'

// Actions
import {
  setLPStatus as setLPStatusAction,
  writeLineProtocolAction,
} from 'src/dataLoaders/actions/dataLoaders'

// Decorator
import {ErrorHandling} from 'src/shared/decorators/errors'

// Types
import {LineProtocolTab} from 'src/types/dataLoaders'
import {AppState} from 'src/types/index'
import {WritePrecision} from '@influxdata/influx'
import {RemoteDataState} from 'src/types'
import {LineProtocolStepProps} from 'src/dataLoaders/components/lineProtocolWizard/LineProtocolWizard'

type OwnProps = LineProtocolStepProps

interface StateProps {
  lineProtocolBody: string
  precision: WritePrecision
  bucket: string
  org: string
}

interface DispatchProps {
  setLPStatus: typeof setLPStatusAction
  writeLineProtocolAction: typeof writeLineProtocolAction
}

type Props = OwnProps & StateProps & DispatchProps

@ErrorHandling
export class LineProtocol extends PureComponent<Props> {
  public componentDidMount() {
    const {setLPStatus} = this.props
    setLPStatus(RemoteDataState.NotStarted)
  }

  public render() {
    return (
      <div className="onboarding-step">
        <Form onSubmit={this.handleSubmit}>
          <FancyScrollbar
            autoHide={true}
            className="wizard-step--scroll-content"
          >
            <div>
              <h3 className="wizard-step--title">Add Data via Line Protocol</h3>
              <h5 className="wizard-step--lp-sub-title">
                Need help writing InfluxDB Line Protocol?{' '}
                <a
                  href="https://v2.docs.influxdata.com/v2.0/write-data/#write-data-in-the-influxdb-ui"
                  target="_blank"
                >
                  See Documentation
                </a>
              </h5>

              {this.content}
            </div>
          </FancyScrollbar>
          <OnboardingButtons autoFocusNext={true} />
        </Form>
      </div>
    )
  }

  private get LineProtocolTabs(): LineProtocolTab[] {
    return [LineProtocolTab.UploadFile, LineProtocolTab.EnterManually]
  }

  private get content(): JSX.Element {
    const {bucket, org} = this.props
    return (
      <LineProtocolTabs
        tabs={this.LineProtocolTabs}
        bucket={bucket}
        org={org}
      />
    )
  }

  private handleSubmit = () => {
    const {onIncrementCurrentStepIndex} = this.props
    this.handleUpload()
    onIncrementCurrentStepIndex()
  }

  private handleUpload = async () => {
    const {bucket, org, lineProtocolBody, precision} = this.props
    this.props.writeLineProtocolAction(org, bucket, lineProtocolBody, precision)
  }
}

const mstp = ({
  dataLoading: {
    dataLoaders: {lineProtocolBody, precision},
    steps: {bucket},
  },
  orgs,
}: AppState): StateProps => {
  return {lineProtocolBody, precision, bucket, org: orgs.org.name}
}

const mdtp: DispatchProps = {
  setLPStatus: setLPStatusAction,
  writeLineProtocolAction,
}

export default connect<StateProps, DispatchProps, OwnProps>(
  mstp,
  mdtp
)(LineProtocol)
