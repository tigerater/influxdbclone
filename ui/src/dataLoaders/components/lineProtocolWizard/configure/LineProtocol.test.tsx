// Libraries
import React from 'react'
import {shallow} from 'enzyme'

// Components
import {LineProtocol} from 'src/dataLoaders/components/lineProtocolWizard/configure/LineProtocol'
import {WritePrecision} from '@influxdata/influx'

const setup = (override = {}) => {
  const props = {
    bucket: 'a',
    org: 'a',
    onClickNext: jest.fn(),
    lineProtocolBody: '',
    precision: WritePrecision.Ns,
    setLPStatus: jest.fn(),
    writeLineProtocolAction: jest.fn(),
    currentStepIndex: 0,
    onIncrementCurrentStepIndex: jest.fn(),
    onDecrementCurrentStepIndex: jest.fn(),
    notify: jest.fn(),
    onExit: jest.fn(),
    ...override,
  }
  const wrapper = shallow(<LineProtocol {...props} />)

  return {wrapper}
}

describe('LineProtocol', () => {
  describe('rendering', () => {
    it('renders!', () => {
      const {wrapper} = setup()
      expect(wrapper.exists()).toBe(true)
    })
  })
})
