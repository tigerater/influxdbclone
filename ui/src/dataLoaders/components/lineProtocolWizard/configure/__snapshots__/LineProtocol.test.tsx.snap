// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineProtocol rendering renders! 1`] = `
<div
  className="onboarding-step"
>
  <Form
    onSubmit={[Function]}
    testID="form-container"
  >
    <FancyScrollbar
      autoHeight={false}
      autoHide={true}
      className="wizard-step--scroll-content"
      hideTracksWhenNotNeeded={true}
      maxHeight={null}
      setScrollTop={[Function]}
      style={Object {}}
    >
      <div>
        <h3
          className="wizard-step--title"
        >
          Add Data via Line Protocol
        </h3>
        <h5
          className="wizard-step--lp-sub-title"
        >
          Need help writing InfluxDB Line Protocol?
           
          <a
            href="https://v2.docs.influxdata.com/v2.0/write-data/#write-data-in-the-influxdb-ui"
            target="_blank"
          >
            See Documentation
          </a>
        </h5>
        <Connect(LineProtocolTabs)
          bucket="a"
          org="a"
          tabs={
            Array [
              "Upload File",
              "Enter Manually",
            ]
          }
        />
      </div>
    </FancyScrollbar>
    <OnboardingButtons
      autoFocusNext={true}
      backButtonText="Previous"
      nextButtonStatus="default"
      nextButtonText="Continue"
      showSkip={false}
      skipButtonText="Skip"
    />
  </Form>
</div>
`;
